components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenu.kt

```diff
@@ -283,3 +291,172 @@ private fun getMaxAvailableHeightToTopAndBottom(anchor: View): Pair<Int, Int> {
+        availableHeightToBottom = availableHeightToBottom,
+        containerViewHeight = containerHeight
+    ))
+}
+
+/**
+ * Infer where and how the PopupWindow should be shown based on the data available in [currentData].
+ * Should be called only once per menu to be shown.
+ *
+ * @param anchor view the PopupWindow will be aligned to.
+ * @param currentData current known data for how the menu should be positioned.
+ *
+ * @return new MenuPositioningData updated to contain the inferred [MenuPlacement]]
+ */
+@VisibleForTesting
+internal fun inferMenuPosition(anchor: View, currentData: MenuPositioningData): MenuPositioningData {
+    // Try to use the preferred orientation, if doesn't fit fallback to the best fit.
+
+    val menuPlacement: MenuPlacement =
+    if (currentData.askedOrientation == DOWN && currentData.fitsDown) {
+        MenuPlacement.AnchoredToTop.Dropdown(anchor)
+    } else if (currentData.askedOrientation == UP && currentData.fitsUp) {
+        MenuPlacement.AnchoredToBottom.Dropdown(anchor)
+    } else {
+        if (!currentData.fitsUp && !currentData.fitsDown) {
+            if (currentData.availableHeightToTop < currentData.availableHeightToBottom) {
+                MenuPlacement.AnchoredToTop.ManualAnchoring(anchor)
+            } else {
+                MenuPlacement.AnchoredToBottom.ManualAnchoring(anchor)
+            }
+        } else {
+            if (currentData.fitsDown) {
+                MenuPlacement.AnchoredToTop.Dropdown(anchor)
+            } else {
+                MenuPlacement.AnchoredToBottom.Dropdown(anchor)
+            }
+        }
+    }
+
+    return currentData.copy(inferredMenuPlacement = menuPlacement)
+}
+
+/**
+ * All data needed for menu positioning.
+ */
+@VisibleForTesting
+internal data class MenuPositioningData(
+    /**
+     * Where and how should the menu be placed in relation to the [MenuPlacement.anchor].
+     */
+    val inferredMenuPlacement: MenuPlacement? = null,
+
+    /**
+     * The orientation asked by users of this class when initializing it.
+     */
+    val askedOrientation: BrowserMenu.Orientation = DOWN,
+
+    /**
+     * Whether the menu fits in the space between [display top, anchor] in a top - down layout.
+     */
+    val fitsUp: Boolean = false,
+
+    /**
+     * Whether the menu fits in the space between [anchor, display top] in a top - down layout.
+     */
+    val fitsDown: Boolean = false,
+
+    /**
+     * Distance between [display top, anchor top margin]. Used for better positioning the menu.
+     */
+    @Px val availableHeightToTop: Int = 0,
+
+    /**
+     * Distance between [display bottom, anchor bottom margin]. Used for better positioning the menu.
+     */
+    @Px val availableHeightToBottom: Int = 0,
+
+    /**
+     * [View#measuredHeight] of the menu. May be bigger than the available screen height.
+     */
+    @Px val containerViewHeight: Int = 0
+)
+
+@VisibleForTesting
+internal sealed class MenuPlacement {
```

Comment:
Nit: same, can we make this its own file?

Reply:
Sure.

Reply:
Wow, the code does look much cleaner, thanks!